3단계 - 자동차 경주

AlbertImKr opened the pull request:
조남재 리뷰어 님 , 안녕하세요 👋

마지막에 기능 통합 때..  입출력관련 및 Service?같은 전달만 하는 역할을 해서 테스트 작성여부에 대하여 고민을 많이 했습니다.

이런 통합 흐름과 입출력 테스트에 대하여 리뷰어 님이 어떻게 생각하시는지 궁금합니다.
-  저는 이미 하위에서 테스트 했고 흐름만 제어하는 메서드에 대하여 테스트 할 필요성이 없다고 생각해요.

AlbertImKr committed:
refactor: 사칙연산자를 enum으로 분리

AlbertImKr committed:
refactor: 정규식으로 input 검증

AlbertImKr committed:
refactor: 사칙연산을 함수로 변경

AlbertImKr committed:
docs: 기능 요구사항 작성

AlbertImKr committed:
test: 값이 1보다 큰 숫자가 아니면 예외를 던지는 테스트 구현

AlbertImKr committed:
feat: 값이 1보다 큰 숫자가 아니면 예외를 던지는 기능 구현

AlbertImKr committed:
docs: 구현 기능 체크

AlbertImKr committed:
test: 자동차는 주어지 숫자만큼 전진 테스트 구현

AlbertImKr committed:
feat: 자동차는 주어지 숫자만큼 전진 기능 구현

AlbertImKr committed:
docs: 구현 기능 체크

AlbertImKr committed:
test: 입력 검증 성공 사례 테스트 구현

AlbertImKr committed:
docs: 기능목록 수정

AlbertImKr committed:
feat: 콘솔에서 입력받는 기능 구현

AlbertImKr committed:
docs: 구현 기능 체크

AlbertImKr committed:
feat: 문자를 콘솔에 츨력 기능 구현

AlbertImKr committed:
docs: 구현 기능 체크

AlbertImKr committed:
docs: 기능목록 수정

AlbertImKr committed:
feat: 0에서 9사이의 랜덤 전진 값을 생성한는 기능 구현

AlbertImKr committed:
docs: 구현 기능 체크

AlbertImKr committed:
docs: 기능목록 수정

AlbertImKr committed:
feat: 기능 통합

AlbertImKr committed:
docs: 구현 기능 체크

namjackson commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,13 @@
+            Operation.MINUS -> currentResult.minus(nextValue)
+            Operation.MULTIPLY -> currentResult.times(nextValue)
+            Operation.DIVIDE -> currentResult.div(nextValue)
+        }
```
Calculator는 어떤 책임을 가질까요? 불필요한 객체는 아닐까요?

실제 연산도 Operation 객체에게 메세지를 보내고, 계산을 위임해보면 어떨까요?
`operation.applyOperation(currentResult: Int, nextValue: Int)`

namjackson commented on src/main/kotlin/calculator/Operation.kt:
```diff
@@ -0,0 +1,15 @@
+
+    companion object {
+        fun fromString(element: String): Operation {
+            values().forEach { if (it.value == element) return it }
```
find라는 확장함수를 참고해보아요!

namjackson commented on src/main/kotlin/game/Car.kt:
```diff
@@ -0,0 +1,9 @@
+package game
+
+class Car(var position: Int) {
+    constructor() : this(0)
```
default value를 활용해보아요!

namjackson commented on src/main/kotlin/game/Car.kt:
```diff
@@ -0,0 +1,9 @@
+package game
+
+class Car(var position: Int) {
```
position을 외부에서 변경할수 있는 구조네요,
외부에서 변경하지 못하게 해보면 어떨까요?

namjackson commented on src/main/kotlin/game/Cars.kt:
```diff
@@ -0,0 +1,18 @@
+            ConsoleTextPrinter.printText(currentPosition)
+        })
+        println()
+    }
```
비즈니스 로직과, 화면 관련 로직이 섞여있어요, 
화면 관련 로직은 분리해보면 어떨까요?

namjackson commented on src/main/kotlin/game/Cars.kt:
```diff
@@ -0,0 +1,18 @@
+
+    fun advance(retryCount: Int) = repeat(retryCount) { advance() }
+
+    constructor(cars: Int) : this(1.rangeTo(cars).map { Car() })
```
생성자를 사용해도 좋지만,
팩토리 함수를 사용해서 가짜 생성자를 활용해보면 어떨까요?
가볍게 읽어보셔도 좋을거같아요!
https://kyucumber.github.io/posts/book/effective-kotlin/effective-kotlin-item-33

namjackson commented on src/main/kotlin/game/Cars.kt:
```diff
@@ -0,0 +1,18 @@
+
+    private fun advance() {
+        cars.forEach(fun(car: Car) {
+            val movement = AdvanceCountGenerator().generateRandomValue()
```
랜덤 관련 기능은 어떻게 테스트 해볼수 있을까요?



(추상화 통해 랜덤 기능을 분리해보아요!)

namjackson commented on src/test/kotlin/game/CarTest.kt:
```diff
@@ -0,0 +1,20 @@
+class CarTest {
+    @ParameterizedTest
+    @CsvSource(value = ["1:2:3", "4:5:9", "0:8:8"], delimiter = ':')
+    fun `자동차는 주어지 숫자만큼 전진`(position: Int, movement: Int, expected: Int) {
```
> 전진하는 조건은 0에서 9 사이에서 무작위 값을 구한 후 무작위 값이 4 이상일 경우이다.

요구사항과 다른거 같아요! 
랜덤값이 4이상일때 이동하는 로직입니다!

namjackson commented on src/main/kotlin/game/RacingCarGame.kt:
```diff
@@ -0,0 +1,21 @@
+
+    fun start() {
+        ConsoleTextPrinter.printText(carCountQuestion)
+        val carCount = ConsoleInputReader.readInput().toInt()
```
Printer, Reader 를 입/출력을 분리하기보단
기능단위로 함수를 분리하는건 어떨까요?

namjackson commented on src/main/kotlin/game/RacingCarGame.kt:
```diff
@@ -0,0 +1,21 @@
+package game
+
+class RacingCarGame() {
```
> 마지막에 기능 통합 때.. 입출력관련 및 Service?같은 전달만 하는 역할을 해서 테스트 작성여부에 대하여 고민을 많이 했습니다.
> 이런 통합 흐름과 입출력 테스트에 대하여 리뷰어 님이 어떻게 생각하시는지 궁금합니다.
> 저는 이미 하위에서 테스트 했고 흐름만 제어하는 메서드에 대하여 테스트 할 필요성이 없다고 생각해요.


이부분은 의견이 다른부분이라 참고만 해주셔도 좋을거 같아요!
단순한 입출력 테스트라면 불필요하다고 생각이 들어요,
하지만, 저는 흐름제어 또한 하나의 비즈니스 로직이라고 생각합니다
테스트 가능하다면 테스트 코드를 작성하면 좋을거같아요!
View관련 부분을 인터페이스로 분리하면다면, 테스트 가능한 구조가 되지않을까요?
RacingCarGame 또한 테스트 가능한 구조로 만들어보면 어떨까요?

namjackson reviewed (changes requested):
안녕하세요 홍광님
3단계 미션 고생하셨습니다! 
몇가지 고민할법한 포인트들 코멘트로 남겼으니, 확인해주세요 😄

AlbertImKr committed:
refactor: Calculator 삭제하고 함수를 Operation 이동

AlbertImKr committed:
refactor: Default 생성자 position 기본값 부여

AlbertImKr committed:
refactor: 생성자 대신 팩토리 함수 구현

AlbertImKr committed:
refactor: 랜덤 전진 전략을 전진 전략 Interface를 상속하는 방식으로 분리

AlbertImKr committed:
test: 테스트 케이스 수정

AlbertImKr committed:
fix: 잘못된 기능 수정

AlbertImKr committed:
test: 자동차는 4이상 9이하 일 때 이동하는 테스트 구현

AlbertImKr committed:
feat: 자동차는 조건이 4이상 9일 때 한칸 전진하는 기능 구현

AlbertImKr committed:
style: 클래스 네이밍 변경

AlbertImKr committed:
refactor: 입력 관련 기능을 기능으로 분리

AlbertImKr committed:
refactor: 출력 관련 기능을 기능으로 분리

AlbertImKr committed:
fix: Input 검증 로직 누락 구현

AlbertImKr committed:
test: Random대신 고정 전진조건 조건생성자를 만들어서 전진 로직 테스트 구현

AlbertImKr committed:
chore: 에러 발생애서 kotest 버전 올림

java.lang.IncompatibleClassChangeError: Found interface kotlin.time.TimeMark, but class was expected

AlbertImKr committed:
test: kotest 사용해서 bahavior 검증

AlbertImKr committed:
test: kotest 사용해서 bahavior 검증

AlbertImKr commented:
리뷰 반영했습니다. 🫡

namjackson commented on src/main/kotlin/calculator/CalculatorValidator.kt:
```diff
@@ -2,17 +2,11 @@ package calculator
-        require(size >= MINIMUM_SIZE) { ErrorMessage.MINIMUM_SIZE_REQUIREMENT_MSG.message }
-        require(size % EVEN_NUMBER_DENOMINATOR != EVEN_NUMBER_REMAINDER) { ErrorMessage.ODD_SIZE_REQUIREMENT_MSG.message }
-    }
+    private val VALID_NUMBER_PATTERN = "^-?\\d+( [+\\-*/] -?\\d+)+\$".toRegex()
```
추가적으로, 연산자에 대한 문자열은 
따로 분리한 연산자 객체에서 관리하면 어떨까요?
연산자가 추가될때마다, 함께 변경되어야하는 구조에서 벗어날수 있을거같아요 😄

namjackson commented on src/main/kotlin/game/Car.kt:
```diff
@@ -0,0 +1,16 @@
+package game
+
+class Car(var position: Int = 0) {
```
var 변수에 접근제어자가 없다면,
외부에서 쉽게 변경할수 있는 구조네요! 
접근제어를 통해 안전하게 관리해보아요!

namjackson commented on src/main/kotlin/game/Cars.kt:
```diff
@@ -0,0 +1,24 @@
+
+    fun advance(retryCount: Int) = repeat(retryCount) {
+        advance()
+        UserMessageDisplay.display(carList)
```
Cars는 비즈니스 로직에 UserMessageDisplay 라는 UI 로직이 여전히 존재하네요,
UI로직을 분리해보아요! 

여러가지 방법이 있을거같아요! advance 함수의 람다 파라미터를 통해 이벤트를 전달하거나,
외부에서  Cars객체의 carList에 접근해서 그려주는 방법이 있을거 같아요!

AlbertImKr replied:
람다 파라미터를 통해 이벤트를 전달하는 법을 알았습니다. 감사합니다.~

namjackson commented on src/main/kotlin/game/RacingCarGame.kt:
```diff
@@ -0,0 +1,16 @@
+class RacingCarGame() {
+
+    fun start() {
+        val carCount = UserInputHandler.askForCarCount()
```
UserInputHandler 또한, 인터페이스로 분리하여,
RacingCarGame 생성시점에 주입하면, 
RacingCarGame 객체도 테스트 가능한 구조가 될거같아요!

namjackson commented on src/test/kotlin/game/FixedMoveConditionGenerator.kt:
```diff
@@ -0,0 +1,5 @@
+package game
+
+class FixedMoveConditionGenerator(private val count: Int) : MoveConditionGenerator {
```
Fake객체 활용 👍 👍

namjackson reviewed (approved):
안녕하세요 홍광님!
피드백 반영 잘해주셨네요!! 
몇가지 추가적으로 코멘트 남겼으니, 확인해주세요!
변경사항은 다음미션 진행하면서, 적용해주셔도 좋을거같아요! 
빠른 미션 진행을 위해 머지하겠습니다!
